### ISSUE TITLE
Open .ksy files directly

### ISSUE THREAD

**Issue by abitrolly** [kaitai-io/kaitai_struct_python_runtime#50]
http://doc.kaitai.io/lang_python.html doesn't specify how exactly to import .ksy files from Python. I would expect something like this.
```python
import kaitai
Gif = kaitai.compile('git.ksy')
g = Gif.from_file('some.gif')
```

**Comment by dgelessus** [comment]
You're correct, there's no built-in support for importing .ksy files directly from Python. The usual workflow is to compile the .ksy files into Python modules using `ksc` (as described in the KS Python notes) and put the compiled Python modules into a location where they can be imported - if you're writing a library, into a package in your library code, or if you're writing a script, in the same directory as the script.

@KOLANICH has written a library called [kaitaiStructCompile.py](https://github.com/kaitaiStructCompile/kaitaiStructCompile.py), which (among other things) lets you import .ksy files as Python modules without manually compiling them first. When you import a KSY using kaitaiStructCompile, it automatically calls `ksc` to generate the Python module, and imports that. But I would only recommend this for code for your own personal use, not for libraries/tools meant to be used by others - otherwise every user of your code needs to have the correct version of `ksc` installed.

**Comment by abitrolly** [comment]
@dgelessus good start. It is technically feasible to ship compiler binary in a Python wheel. I made a fork in https://github.com/abitrolly/ksykaitai with a roadmap, but no ETA how it could be done.

**Comment by abitrolly** [comment]
Two hours later still can't get it running.

```python
$ python ksykaitai.py
Traceback (most recent call last):
  File "ksykaitai.py", line 18, in <module>
    compile('../kaitai_struct_visualizer/squashfs_superblock.ksy')
  File "ksykaitai.py", line 9, in compile
    backend = clibackend.init(ICompilerModule,
  File "/home/anatoli/h/snapsquash/ksykaitai/kaitaiStructCompile/backend/cmdline.py", line 87, in init
    class CLIPrefsStorage(ICompilerModule.IPrefsStorage):
AttributeError: type object 'ICompiler' has no attribute 'IPrefsStorage'
```

7566f4e committed https://github.com/abitrolly/ksykaitai

@KOLANICH I committed backed to repo to avoid installation hassle, and yet it doesn't work.

**Comment by KOLANICH** [comment]
>type object 'ICompiler' has no attribute 'IPrefsStorage'

[`IPrefsStorage`](https://github.com/kaitaiStructCompile/kaitaiStructCompile.py/blob/master/kaitaiStructCompile/ICompiler.py#L71) is in `ICompiler` module, not class (the module also contains an eponymous class).

**Comment by KOLANICH** [comment]
>It is technically feasible to ship compiler binary in a Python wheel.

It is an extremily bad idea. It depends on JVM. You are not going to also ship JVM as a wheel, are you?

So it should be handled either manually, or by a system package manager.

**Comment by abitrolly** [comment]
> It is an extremily bad idea. KSC depends on JVM. You are not going to also ship JVM as a wheel, are you?

Going. Exactly. :D https://github.com/dgiagio/warp#quickstart-with-java

> IPrefsStorage is in ICompiler module, not class (the module also contains an eponymous class).

So how to import ICompiler module? The class import in `__init__.py` shadows it.

**Comment by KOLANICH** [comment]
>Going. Exactly

🤦‍♂️


>So how to import ICompiler module? The class import in __init__.py shadows it.

1. yes, it shadows. Probably I should rename that module.
2. It is not meant to be imported. It is injected into selected backends. The backends are discovered from entry points and usually are not needed to be constructed manually.

**Comment by abitrolly** [comment]
> Going. Exactly.

https://github.com/kaitai-io/kaitai_struct/issues/807

> yes, it shadows. Probably I should rename that module.

Answering your style - you should remove that module and the module with exception. Just saying. :speak_no_evil: 

Anyway, fixed in https://github.com/abitrolly/ksykaitai/commit/293c892883d6f5257902f7053012d170f95d38ab

Right now the problem I am dealing with is the path to compiler binary. The Python lib expects it to be a directory structure, and I am running the compiler using a wrapper script (which crates secure environment).

**Comment by KOLANICH** [comment]
>The Python lib expects it to be a directory structure, and I am running the compiler using a wrapper script (which crates secure environment).

You may be able to set `KAITAI_STRUCT_ROOT` to point inside that environment (is it still a dir?)

**Comment by abitrolly** [comment]
With the latest commit I successfully compile my .ksy from Python (after downloading compiler, unpacking to current dir and stripping version suffix). The code works! Although currently I am not sure how it is better than a direct subprocess call. :D

**Comment by abitrolly** [comment]
Added automatic import of generated file.

Also committed data to be used as test and proof of concept. Except compiler, because it is too big. The rest of the work is removing hardcoded filenames and creating a wheel with compiler bundled for amd64 Linux and accessing bundled data files with https://importlib-resources.readthedocs.io/en/latest/index.html
